lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3896,6 +3896,53 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     }
   }
 
+  assert(!toType->isAny() && "casts to 'Any' should've been handled above");
+  assert(!toType->isAnyObject() &&
+         "casts to 'AnyObject' should've been handled above");
+
+  // A cast from a function type to an existential type (except `Any`)
+  // or an archetype type (with constraints) cannot succeed
+  auto toArchetypeType = toType->is<ArchetypeType>();
+  auto fromFunctionType = fromType->is<FunctionType>();
+  auto toExistentialType = toType->isAnyExistentialType();
+
+  auto toConstrainedArchetype = false;
+  if (toArchetypeType) {
+    auto archetype = toType->castTo<ArchetypeType>();
+    toConstrainedArchetype = archetype && !archetype->getConformsTo().empty();
```

Comment:
Sorry, what I meant here is that the archetype cannot be null, because toArchetypeType true implies toType is an ArchetypeType. That's why I said to use castTo<> (which trails on failure) instead of getAs<> (which returns null). So you don't need the null check either.

Do you mind fixing that in a follow-up commit?

Reply:
Ah ok, I have created another PR to remove it: #23167